pkg/controller/mysqlcluster/mysqlcluster_controller_test.go

```diff
@@ -17,65 +16,298 @@ limitations under the License.
-	mysqlv1alpha1 "github.com/presslabs/mysql-operator/pkg/apis/mysql/v1alpha1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+
 	"golang.org/x/net/context"
-	appsv1 "k8s.io/api/apps/v1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	policy "k8s.io/api/policy/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	api "github.com/presslabs/mysql-operator/pkg/apis/mysql/v1alpha1"
 )
 
-var c client.Client
+type clusterComponents []runtime.Object
+
+const timeout = time.Second * 2
+
+var _ = Describe("MysqlCluster controller", func() {
+	var (
+		// channel for incoming reconcile requests
+		requests chan reconcile.Request
+		// stop channel for controller manager
+		stop chan struct{}
+		// controller k8s client
+		c client.Client
+	)
+
+	BeforeEach(func() {
+		var recFn reconcile.Reconciler
+
+		mgr, err := manager.New(cfg, manager.Options{})
+		Expect(err).NotTo(HaveOccurred())
+		c = mgr.GetClient()
+
+		recFn, requests = SetupTestReconcile(newReconciler(mgr))
+		Expect(add(mgr, recFn)).To(Succeed())
+
+		stop = StartTestManager(mgr)
+	})
+
+	AfterEach(func() {
+		close(stop)
+	})
+
+	Describe("when creating a new mysql cluster", func() {
+		var (
+			expectedRequest reconcile.Request
+			cluster         *api.MysqlCluster
+			secret          *core.Secret
+			components      clusterComponents
+		)
+
+		BeforeEach(func() {
+			name := fmt.Sprintf("cluster-%d", rand.Int31())
+			ns := "default"
+
+			expectedRequest = reconcile.Request{
+				NamespacedName: types.NamespacedName{Name: name, Namespace: ns},
+			}
+
+			secret = &core.Secret{
+				ObjectMeta: metav1.ObjectMeta{Name: "the-secret", Namespace: ns},
+				StringData: map[string]string{
+					"ROOT_PASSWORD": "this-is-secret",
+				},
+			}
+
+			cluster = &api.MysqlCluster{
+				ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: ns},
+				Spec: api.MysqlClusterSpec{
+					Replicas:   1,
+					SecretName: secret.Name,
+				},
+			}
+
+			components = []runtime.Object{
+				&apps.StatefulSet{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.StatefulSet),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.HeadlessSVC),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.MasterService),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.HealthyNodesService),
```

Comment:
Do not call `GetNameForResource`. Hardcode the expected name in test, so you know if something is wrong with `GetNameForResource`.